compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java

```diff
@@ -5084,16 +5085,14 @@ private ParserRuleContext getNextRuleForVarName() {
     private ParserRuleContext getNextRuleForSemicolon(int nextLookahead) {
         STToken nextToken;
         ParserRuleContext parentCtx = getParentContext();
-        if (parentCtx == ParserRuleContext.EXTERNAL_FUNC_BODY) {
-            endContext(); // end external func-body
-            return getNextRuleForSemicolon(nextLookahead);
-        } else if (parentCtx == ParserRuleContext.QUERY_EXPRESSION) {
-            endContext(); // end expression
+        if (parentCtx == ParserRuleContext.EXTERNAL_FUNC_BODY || parentCtx == ParserRuleContext.ON_FAIL_CHECK
+                || parentCtx == ParserRuleContext.QUERY_EXPRESSION) {
+            endContext();
             return getNextRuleForSemicolon(nextLookahead);
         } else if (isExpressionContext(parentCtx)) {
             // A semicolon after an expression also means its an end of a statement/field, Hence pop the ctx.
-            endContext(); // end statement
-            return ParserRuleContext.STATEMENT;
+            endContext();
+            return getNextRuleForSemicolon(nextLookahead);
```

Comment:
Do we need to change these?

1. We need to change the next rule for the double arrow first.
https://github.com/ballerina-platform/ballerina-lang/blob/08736f60516c387939ca373b75d282090de03374/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java#L3800

Above should now be `ParserRuleContext.EXPRESSION`

Reply:
2. We have to remove this condition from here
https://github.com/ballerina-platform/ballerina-lang/blob/08736f60516c387939ca373b75d282090de03374/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java#L3582-L3585

Reply:
3. We have to add a new if condition for `ParserRuleContext.ON_FAIL_CHECK` here and give the next rule as
`ParserRuleContext.ON_FAIL_CHECK_END` which would pop both `ON_FAIL_CHECK` and `CHECKING_EXPRESSION` contexts and point next rule as `EXPRESSION_RHS`
 
https://github.com/ballerina-platform/ballerina-lang/blob/08736f60516c387939ca373b75d282090de03374/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java#L3081

Reply:
Updated [8a51481](https://github.com/ballerina-platform/ballerina-lang/pull/41098/commits/8a514812a9aa20e6f470e046c943e55c47d782f5)